src/webutil/tableState.js

```diff
@@ -0,0 +1,244 @@
+      />
+  Filtering:
+      <input type="text" onChange={(event) => {
+        //Do not include an Event object from the DOM in the filter function as Event objects do not persist.
+        const value = event.target.value;
+        ts.createOrUpdateFilterFn(
+          "filterEventName",
+          (ledgerEvent) => ledgerEvent.action.type.replace("_", " ").toLowerCase().includes(value.toLowerCase())
+        )    
+      }}/>
+  Sorting:
+      <TableSortLabel
+          active={ts.sortName === DATE_SORT.name}
+          direction={
+            ts.sortName === DATE_SORT.name ? ts.sortOrder : DEFAULT_SORT
+          }
+          onClick={() => ts.setSortFn(DATE_SORT.name, DATE_SORT.fn)}
+      >Date</TableSortLabel>
+ */
+export type TableState<T> = {|
+  /**
+    A stateful representation of the current page data.
+   */
+  +currentPage: $ReadOnlyArray<T>,
+  /**
+    A stateful representation of the length of the filtered unpaginated data.
+   */
+  +length: number,
+  /**
+    A stateful representation of the current page index.
+   */
+  +pageIndex: number,
+  /**
+    A stateful representation of the number of rows per page.
+   */
+  +rowsPerPage: number,
+  /**
+    A stateful representation of the current sort name (the last sort name passed to setSortFn).
+   */
+  +sortName: string,
+  /**
+    A stateful representation of the current order, represented by "asc" or "desc".
+   */
+  +sortOrder: SortOrderTypes,
+  /**
+    A function that changes the currentPage to represent the indicated index.
+   */
+  +setPageIndex: (number) => void,
+  /**
+    A function that sets the rows per page and updates the pageIndex to an equivalent place.
+   */
+  +setRowsPerPage: (number) => void,
+  /**
+    A function that adds a filter to the map of filters or overwrites an existing filter that matches the key.
+    Every filter stored will be applied together as an AND operation (results will only include elements
+    that pass all filters).
+    Sets pageIndex to 0.
+
+    string: A key that represents the filter in the map.
+    function?: A function that returns false if an element should be excluded from the result. Null to remove the filter at the key.
+   */
+  +createOrUpdateFilterFn: (string, null | ((T) => boolean)) => void,
+  /**
+    A function that sorts the data. Current pageIndex is maintained.
+
+    string: A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    function?: A function that returns the property of the element that should be used in the sort comparator.
+   */
+  +setSortFn: (string, PluckFn<T>) => void,
+|};
+
+type FilterFns<T> = {[string]: null | ((T) => boolean)};
+
+/**
+  A builder function for initializing a new TableState for use in a ReactNode.
+  It encapsulates Sort, Pagination, and Filtering.
+ */
+export function useTableState<T>(
+  // The base array. It should ideally be memoized before being passed in.
```

Comment:
What do you mean by memoizing the array?

Reply:
That it should be referentially equivalent on a re-render, wrapped in a react memo/state. Is there a better way to say this?